libs/SalesforceSDK/src/com/salesforce/androidsdk/push/PushNotificationDecryptor.java

```diff
@@ -0,0 +1,133 @@
+ * list of conditions and the following disclaimer.
+ * - Redistributions in binary form must reproduce the above copyright notice,
+ * this list of conditions and the following disclaimer in the documentation
+ * and/or other materials provided with the distribution.
+ * - Neither the name of salesforce.com, inc. nor the names of its contributors
+ * may be used to endorse or promote products derived from this software without
+ * specific prior written permission of salesforce.com, inc.
+ * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
+ * AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+ * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
+ * ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT OWNER OR CONTRIBUTORS BE
+ * LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
+ * CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
+ * SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
+ * INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
+ * CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+ * ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+ * POSSIBILITY OF SUCH DAMAGE.
+ */
+package com.salesforce.androidsdk.push;
+
+import android.text.TextUtils;
+import android.util.Base64;
+
+import com.google.firebase.messaging.RemoteMessage;
+import com.salesforce.androidsdk.analytics.security.Encryptor;
+import com.salesforce.androidsdk.app.SalesforceSDKManager;
+import com.salesforce.androidsdk.security.KeyStoreWrapper;
+import com.salesforce.androidsdk.security.SalesforceKeyGenerator;
+
+import java.security.PrivateKey;
+import java.util.Map;
+
+/**
+ * This class processes incoming push notifications and passes them along to the app.
+ * It decrypts the incoming notification if encryption of push notifications is enabled.
+ *
+ * @author bhariharan
+ */
+class PushNotificationDecryptor {
+
+    private static String CONTENT_KEY = "content";
+    private static PushNotificationDecryptor INSTANCE = new PushNotificationDecryptor();
+
+    /**
+     * Returns a singleton instance of this class.
+     *
+     * @return Singleton instance of this class.
+     */
+    static synchronized PushNotificationDecryptor getInstance() {
+        return INSTANCE;
+    }
+
+    void onPushMessageReceived(RemoteMessage message) {
+        final Map<String, String> data = processNotificationPayload(message.getData());
+        if (data != null && SalesforceSDKManager.hasInstance()) {
+            final PushNotificationInterface pnInterface = SalesforceSDKManager.getInstance().getPushNotificationReceiver();
+            if (pnInterface != null) {
+                pnInterface.onPushMessageReceived(data);
+            }
+        }
+    }
+
+    private Map<String, String> processNotificationPayload(Map<String, String> data) {
```

Comment:
Processes the incoming notification - checks if it's encrypted or not based on the fields available.